lib/sequelize.js

```diff
@@ -614,24 +614,30 @@ class Sequelize {
         [sql, bindParameters] = this.dialect.Query.formatBindParameters(sql, options.bind, this.options.dialect);
       }
 
+      const checkTransaction = () => {
+        if (options.transaction && options.transaction.finished && !options.completesTransaction) {
+          const error = new Error(`${options.transaction.finished} has been called on this transaction(${options.transaction.id}), you can no longer use it. (The rejected query is attached as the 'sql' property of this error)`);
+          error.sql = sql;
+          throw error;
+        }
+      };
+
       const retryOptions = Object.assign({}, this.options.retry, options.retry || {});
 
       return Promise.resolve(retry(() => Promise.try(() => {
         if (options.transaction === undefined && Sequelize._cls) {
           options.transaction = Sequelize._cls.get('transaction');
         }
-        if (options.transaction && options.transaction.finished) {
-          const error = new Error(`${options.transaction.finished} has been called on this transaction(${options.transaction.id}), you can no longer use it. (The rejected query is attached as the 'sql' property of this error)`);
-          error.sql = sql;
-          throw error;
-        }
+
+        checkTransaction();
 
         return options.transaction
           ? options.transaction.connection
           : this.connectionManager.getConnection(options);
       }).then(connection => {
         const query = new this.dialect.Query(connection, this, options);
         return this.runHooks('beforeQuery', options, query)
+          .tap(() => checkTransaction())
```

Comment:
Why `checkTransaction` is called here? To check deadlocks caused by `beforeQuery`?

Also just use `then` here, we aren't using return value in promise chain

Reply:
Correct.  I wanted to move this check as close to the actual execution of the query as possible.  

If a commit/rollback/deadlock happens _during_ `getConnection()` or `beforeQuery`, I don't want to mistakenly run a query outside of a transaction.  

Because `beforeQuery` and `getConnection()` can both be asynchronous, I'm actually more concerned about something in the background finishing the transaction, rather than `beforeQuery` committing the transaction itself.

I'll change this to a `then`.